clang/lib/CodeGen/CGExpr.cpp

```diff
@@ -956,42 +956,72 @@ static llvm::Value *getArrayIndexingBound(CodeGenFunction &CGF,
   return nullptr;
 }
 
-const Expr *
+namespace {
+
+struct MemberExprBaseVisitor
+    : public StmtVisitor<MemberExprBaseVisitor, Expr *> {
+  MemberExprBaseVisitor() = default;
+
+  //===--------------------------------------------------------------------===//
+  //                            Visitor Methods
+  //===--------------------------------------------------------------------===//
+
+  Expr *Visit(Expr *E) {
+    return StmtVisitor<MemberExprBaseVisitor, Expr *>::Visit(E);
+  }
+
+  Expr *VisitArraySubscriptExpr(ArraySubscriptExpr *E) {
+    return Visit(E->getBase());
+  }
+  Expr *VisitCastExpr(CastExpr *E) { return Visit(E->getSubExpr()); }
+  Expr *VisitDeclRefExpr(DeclRefExpr *E) { return E; }
+  Expr *VisitMemberExpr(MemberExpr *E) { return Visit(E->getBase()); }
+  Expr *VisitParenExpr(ParenExpr *E) { return Visit(E->getSubExpr()); }
+  Expr *VisitUnaryOperator(UnaryOperator *E) { return Visit(E->getSubExpr()); }
```

Comment:
Do we need to handle binary operators for pointer-to-data-members?

Reply:
Do you mean something using pointer arithmetic? Possibly, though it might be tricky. Do you have an example?

Reply:
I was thinking of something along the lines of:
```
struct S {
  int x, y;
  int blah[] __counted_by(x);
} s;

int main() {
  int (S::*IHatePMDs)[] = &S::blah;
  (s.*IHatePMDs)[0] = 12;
}
```
https://godbolt.org/z/PfzbGd34e

You know... normal everyday code that normal, well-adjusted people think of! Err... ;-)

Reply:
<response voice="Bjarne">You should obviously use a `std::vector!`</response> ;-)

Ooohhh....my head. Okay. Hmm... Technically, flexible array members aren't valid in C++. So could we just smack any developer who tries to do this? ... Please? ;-)

I'll add a check for something like this.

Reply:
We support FAM in C++ as an extension unfortunately. That said, if this is a PITA to support, I'm happy enough with a test demonstrating we don't crash and a FIXME comment saying "support this if a real person ever comes up with something like it." :-D

Reply:
Ah! I made this attribute a C-only feature, so it's ignored in C++. *whew* I'll add the comment though.

Reply:
Excellent!